### ISSUE TITLE
User identity associated with the RP should not be sent to the IDP before explicit consent is gathered

### ISSUE THREAD

**Issue by bvandersloot-mozilla** [fedidcg/FedCM#230]
There is a potential attack that discloses that a user visited an RP to an IDP without explicit consent.

1. RP creates a credential, specifying the IDP in the `federated` parameter and `required` mode.
2. The browser begins executing the [create-credential](https://fedidcg.github.io/FedCM/#create-credential) algorithm.
3. The browser [fetches the manifest](https://fedidcg.github.io/FedCM/#fetch-the-manifest) from the IDP. As specified, this sends a Referer header with value of the RP.
4. The IDP returns a unique account-list url in the manifest and remembers the tuple (account-list url, RP from referer).
5. The browser requests the account list via the url in the manifest, sending its IDP cookie.
6. The IDP checks for the account-list url among its tuples. Upon finding one, it creates a new tuple (IDP cookie from account-list request, Referer header from manifest request).
7. The IDP and RP have won.

There are two features that enable this attack: arbitrary URLs allowed in the manifest and the Referer header being sent with the manifest. Neither is explicitly permitted, but since they are explicitly forbidden in other API call in the spec I assume that they are allowed behaviors. Let me know if I'm misreading this at all!

**Comment by npm1** [comment]
I don't think this is quite correct. Per this table the referrer would not be sent on the manifest (fedcm.json) https://github.com/fedidcg/FedCM/blob/main/explainer.md#endpoint-properties

**Comment by dj2** [comment]
The step `3` above isn't correct, but isn't easy to determine from the spec at the moment. We should update the `fedcm.json` section to specify it is queried _without_ `referer` or `client_id`. The `fedcm.json` file should not be associated to an RP.

**Comment by dj2** [comment]
We did clarify this in the explainer as @npm1 but it hasn't been backported to the spec yet.

**Comment by samuelgoto** [comment]
That's quite an interesting attack @bvandersloot-mozilla , thanks for pointing it out!

While I agree with @npm1 and @dj2 that the **specific** attack you suggested is not viable (we need to fix the spec, PR on the way), I think there we should use this as an opportunity to find attacks that are akin to it.

For example, I think the the following attack is one that we wouldn't be able to mitigate right now:

```js
const cred = await navigator.credentials.get({
  federated: {
    providers: [{
      url: `https://idp.example/${window.location.href}`
    }]
  }
});
// The following will fetch the manifest JSON file, which will know the origin of the RP :(
const {idToken} = cred.login(); 
```

**Comment by bvandersloot-mozilla** [comment]
Thanks for the info @dj2 and @npm1. Good to see this doesn't get through as diagrammed.

@samuelgoto : That is a good catch. This is related to #231. Both are something like [link decoration](https://privacycg.github.io/nav-tracking-mitigations/#link-decoration) of the IDP API endpoints.

**Comment by bvandersloot-mozilla** [comment]
I agree that there are two vectors here, and I think that link decoration is the more concerning one. Timing attacks are definitely the more challenging vector to prevent. Until people begin to talk about removing third-party _requests_, I'm happy to punt on those :)

I think some pre-declaration of the API is a really useful tool. I worry about frequent or just-in-time uses of `FederatedCredential.registerIDP("https://idp.example/${nonce}/")`. For example, an IDP could define a "Log in with another account" button that causes a popup that calls the function then proceeds with the API, disclosing the user cookie. Could this API be restricted from call-to-use or call-to-call?

Another option is to have the account check occur after a user consent is gathered. This would require new structure in the browser because right now the only way to get the account state is to go through the IDP API. This would bind some permission to the RP and IDP pair that permits these requests. This starts to feel a little Storage Access API-like, but hopefully I've conveyed the picture. I think I prefer pre-declaration of the API endpoint though.

**Comment by cbiesinger** [comment]
The downside of having account checks after user consent is that it would require two dialogs (first consent, then account chooser), which is a much worse UX. It would be much better if we could find a viable way that did not require that.

**Comment by asr-enid** [comment]
Without having though this through in depth - why not simply disallow path / fragments / query parts for the IDP URL? The only downside I see is that you can't run multiple IDPs on the same eTLD+1, but that might be an issue for some scenarios.

**Comment by dj2** [comment]
We had something like this originally where the URL was the IDP origin and the manifest was at `.well-known/fedcm.json`. We hard from some IDPs that this had issues for their setup and were requested to take the path into account for the IDP. So we switched to the current model which is a path provided by the RP and `fedcm.json` lives at that path.

**Comment by bvandersloot-mozilla** [comment]
> We had something like this originally where the URL was the IDP origin and the manifest was at `.well-known/fedcm.json`. We hard from some IDPs that this had issues for their setup and were requested to take the path into account for the IDP. So we switched to the current model which is a path provided by the RP and `fedcm.json` lives at that path.

@dj2: Is there a thread or minutes for this? I'd like to know a little more about the use case and if there is a way to handle it less generally.

To make something explicit: an IDP could still smuggle the RDP in on the subdomain unless we restrict the path to only the eTLD+1 and well known URL as @asr-enid implied.

**Comment by samuelgoto** [comment]
> @dj2: Is there a thread or minutes for this? I'd like to know a little more about the use case and if there is a way to handle it less generally.

Here:

https://github.com/fedidcg/FedCM/issues/160#issuecomment-1022773495

**Comment by yi-gu** [comment]
For most consumer use cases, we'd expect a very small number(N) of IDPs per eTLD+1. While having the manifest under root (N=1) is not very extensible, @samuelgoto came up with an alternative proposal that tries to address this issue via an extra layer of indirection.

### RP side
The RP API remains the same:
```
navigator.credentials.get({
  federated: {
    providers: [{
      "url": "https://foo.idp.example/sub",
      "clientId": "1234"
    }]
  }
});
```
### IDP side
#### Unchanged
The IDP has the same manifest in `https://foo.idp.example/sub/fedcm.json` as they do today
```
{
  "accounts_endpoint": …,
  "client_metadata_endpoint": …,
  "id_token_endpoint": …,
  "revocation_endpoint": …
  ...
}
```
#### Changed
In addition, in this formulation, the IDP adds (a one time cost) an extra .well-known file under the root of eTLD+1 (`https://idp.example/.well-known/fedcm.json`) and point to the plausible IDP manifests:
```
{
  "foo.idp.example/sub": "https://foo.idp.example/sub/fedcm.json",
  "bar.idp.example": "https://bar.idp.example/fedcm.json",
  …
}
```
The idea here is that a browser can check that the list of IDPs in the global .well-known file is sufficiently small to represent IDPs (say, a handful per company) but not RPs (say, thousands). 

**The proposal is to start with N=1**, and make this larger as we learn that companies have more than 1 IDP.

### Notes
- Technically speaking, the manifest attack problem is not completely resolved. The IDP can still correlate the user information with the N urls (RP related) listed in the .well-known file. However this should not be a big concern because:
  - N should be sufficiently small so it does not scale
  - .well-known is public
  - Browser can monitor the size of the url list to observe potential abuse
- One more network request to the IDP may be a problem for network with high latency. This could be mitigated by (not mutually exclusive)
  - Fetching the global and the local manifest in parallel and only proceed with the credentialed fetch when the urls from both responses match
  - Caching the two manifests (by N days specified by IDP during fetch)

**Comment by samuelgoto** [comment]
@bvandersloot-mozilla above is another thought that occurred to us as an alternative to the registration alternative we described [here](https://github.com/fedidcg/FedCM/issues/230#issuecomment-1067059128).

The advantage of this approach is that it: (a) it only requires a one-time configuration cost to add the global .well-known file while (b) fetching the .well-known file dynamically rather than registering it.

We think the [alternative](https://github.com/fedidcg/FedCM/issues/230#issuecomment-1067059128) would be hard because it could get easily out of sync (race conditions, e.g. do we fail if we don't see anything registered before hand?) or attacked (e.g. can the RP collude with the IDP to make the IDP "register" moments before we use it? and if so, would we have to impose some sort of "old enough" criteria?). 

WDYT?

**Comment by bvandersloot-mozilla** [comment]
@samuelgoto 

I think I have two points to follow up on here:

1. Given the diversity of the web and microsoftonline.com's multi-tenancy use case, I don't think there is an N that makes sense here. Presumably for microsoftonline.com to work with this, N would have to be as large as the number of Azure AD tenants in existence and be updated every time someone registers a new tenant? I don't know the exact number here, but I registered a tenant for a side project in the last year to enable "Sign in with Microsoft" on a domain I owned. So I assume N would have to be large if I could have already increased it personally. 
1. I don't think that moving from a Javascript registration of the manifest .well-known location (`FederatedCredential.registerIDP("https://idp.example");`) to the scheme @yi-gu describes [here](https://github.com/fedidcg/FedCM/issues/230#issuecomment-1081282938) completely solves the attack we are worried about. The Client-side Javascript of the IDP could notify the server immediately before it creates a credential that a request is coming with a particular provider URL. This would probably rely on fingerprinting though, so I suppose that is an improvement.

**Comment by cbiesinger** [comment]
@bvandersloot-mozilla could you elaborate on your attack in 2) a little bit? How would fingerprinting work?

One thing we're already aware of is this: since FedCM makes a client metadata request (with client ID & referrer) at about the same time as the account list request (with credentials), the server could try to correlate the two requests based on time and IP address and therefore know that a user was visiting the site even when they decline login. Is the attack you're thinking of more or less the same as this one?

**Comment by bvandersloot-mozilla** [comment]
> @bvandersloot-mozilla could you elaborate on your attack in 2) a little bit? How would fingerprinting work?
> 
> One thing we're already aware of is this: since FedCM makes a client metadata request (with client ID & referrer) at about the same time as the account list request (with credentials), the server could try to correlate the two requests based on time and IP address and therefore know that a user was visiting the site even when they decline login. Is the attack you're thinking of more or less the same as this one?

Almost identical! It is a very similar attack, but a little different mechanism. I was thinking of RP Javascript just sending the URI over to the IDP with the Fetch API immediately before creating a credential with the IDP. I just didn't see where in the spec the Metadata endpoint was ever hit, so I didn't want to rely on that.

While fingerprinting is hard to solve entirely, we can try to reduce its feasibility and precision. But having a same-moment pair of requests from the same client containing the information we don't want correlated can lead to a very confident link.

**Comment by cbiesinger** [comment]
Hmm the spec does seem unclear when the client metadata endpoint gets called. The idea is that it is fetched after the manifest, either concurrently or in sequence with the account list endpoint.

The attack is not 100% precise since the server only has the IP address and time to correlate, but yeah, we are aware that this is something we need to solve.

**Comment by bvandersloot-mozilla** [comment]
This seems fundamental to a one-click account chooser. Would it be UX breaking to put a IDP provider chooser before the account chooser?

**Comment by samuelgoto** [comment]
> This seems fundamental to a one-click account chooser. 

One option to fix this that we are exploring is what we've been calling the "push model". In short, in the push model, the IDPs would push the accounts to the browser ahead of time, rather than be pulled dynamically. We think that pushing (as opposed to pulling) would mitigate all of these fingerprinting/timing problems. 

Because we are expecting the push model to be super complex/involved, we are also trying to understand how important are these timing attacks, compared to other options an attacker would have (e.g. IP, other fingerprinting surface and bounce tracking).

Any thoughts on whether the "push model" would work?

> Would it be UX breaking to put a IDP provider chooser before the account chooser?

Yeah, that could work, but would lead to awkward UX. One of the questions that occurred to us is whether the "IDP chooser" should be in the content area (e.g. only allow the API to be called behind a user gesture) or the browser UI.

**Comment by bvandersloot-mozilla** [comment]
> Any thoughts on whether the "push model" would work?

My vision, based on my own priors and what you've said of the push model:

1. User logs into the IDP as a first party.
2. IDP calls a browser API to set up the account at the IDP in the browser.
3. Some time later the user presses an RP button that says Log In With X which calls a browser API.
4. The browser searches through its list of accounts that match all X the RP supports
5. The browser presents an account chooser to the user, and includes a link that navigates to the login page of each IDP with no accounts found.
6. User selects an IDP, Auth proceeds using storage, user is logged in.
7. Success

This does incur browser state for each user IDP account. These are rendered legible to the browser, and is more complicated in the browser. However I would be optimistic about the push model from a privacy perspective. There is no communication with the IDP between when top-level authentication is finished and when the user selects them for a given RP.

> Yeah, that could work, but would lead to awkward UX. One of the questions that occurred to us is whether the "IDP chooser" should be in the content area (e.g. only allow the API to be called behind a user gesture) or the browser UI.
 
My first instinct is to make the IDP chooser in the browser UI, and if you are concerned about too much browser UI you could allow the RP to manage account choice in the content area . The user-RP privacy is at risk when an IDP is contacted, so I'd prefer to have that be gated by a browser-user interaction.


> we are also trying to understand how important are these timing attacks, compared to other options an attacker would have (e.g. IP, other fingerprinting surface and bounce tracking).

w.r.t. IP and other fingerprinting: those are noisy mechanisms that, with work, may be brought back to k-anonymous for useful values of k. The attack here leaks a k=1 identifier (cookie) under reasonable conditions (one login per IP per O(RTT)). As solutions are used to make IP fingerprinting less practical, the privacy gains against all IP fingerprinting would be constrained by this attack because of how it allows precise timing to be added to the mechanism.

**Comment by samuelgoto** [comment]
> My vision, based on my own priors and what you've said of the push model:
> ...
> This does incur browser state for each user IDP account. These are rendered legible to the browser, and is more complicated in the browser. However I would be optimistic about the push model from a privacy perspective. There is no communication with the IDP between when top-level authentication is finished and when the user selects them for a given RP.

Yeah, that sequence of steps matches more or less what I had in mind too, and the privacy/security trade-offs. It is substantially more complicated (for IDPs), so we've avoided it at first.

It seems like we are in agreement that The Push Model would sufficiently mitigate the timing attack problem. Ack?

> My first instinct is to make the IDP chooser in the browser UI, and if you are concerned about too much browser UI you could allow the RP to manage account choice in the content area . The user-RP privacy is at risk when an IDP is contacted, so I'd prefer to have that be gated by a browser-user interaction.

Do you think that gating behind user gesture is a sufficient mitigation? That is, we kept the "pull model" but gated it behind a browser-user interaction, do you think that would decrease the attack surface enough to mitigate it?

**Comment by gffletch** [comment]
There are many cases where the user doesn't go to the IDP first and in those cases today, the RP redirects the user to the IDP. So in that context there would be no chance for the IDP to "pre-register" the user with the browser. Using a model where the user is required to register their accounts with the browser would work better.

Basically, if the user goes to an RP and the RP starts the FedCM process and the browser has not user accounts registered, the browser asks the user to first identify an IDP they want to use and the browser walks the user through getting the account/identity registered with the browser. These identities then can be used in the future without requiring an IDP chooser. Just build the new IDP identity registration into the browser identity UX. This basically becomes an "identity wallet" within the browser.

**Comment by bvandersloot-mozilla** [comment]
@gffletch : 

There is a lighter weight middle-ground here. When the user indicates which IDP they want to use, the browser could open a window/tab that is the login page for that IDP, which on completion would call the "pre-registration" API, which could notify any pending IDP selections.

> Just build the new IDP identity registration into the browser identity UX
The word "just" here is doing a lot of work :)

**Comment by timcappalli** [comment]
You're ultimately proposing that federated accounts get linked to the browser after the user signs into the IdP. What will prevent the browser, browser owner, or sync fabric operator from abusing this global linkage?

**Comment by bvandersloot-mozilla** [comment]
> From other parts of the FedCM proposal, this likely would be linked beyond just the single browser instance.
Interesting, I hadn't put too much thought into cross-browser sync. I was picturing a platform, not browser level feature. I would be in support of language saying that the IDP registration state must not be synced beyond the current device.

> What will prevent the browser, browser owner, or sync fabric operator from abusing this global linkage?
I agree that browser abuse potential is high under this "push" model, and needs to be something we design with in mind. In my personal threat model this is a lost battle, however. I assume the browser owner is aware of any logins I perform, including to IDPs, and it is up to their privacy policy to control whether that information is collected and how it is handled if so. Can you elaborate your position? I'm open to arguments that the degree logins are made legible to the browser requires tighter controls.

**Comment by timcappalli** [comment]
> I assume the browser owner is aware of any logins I perform

This is concerning. This ultimately leads to a world where one company blocks everyone else out but has exclusive access to a superset of data. We recently saw this happen on mobile platforms with platform changes regarding device advertising identifiers.

**Comment by gffletch** [comment]
> Just build the new IDP identity registration into the browser identity UX
> The word "just" here is doing a lot of work :)

I would argue that FedCM (in it's current proposal) is requiring a lot of work across the identity eco-system :)

**Comment by samuelgoto** [comment]
> I would argue that FedCM (in it's current proposal) is requiring a lot of work across the identity eco-system :)

Yeah, agreed, more specifically, IDPs. While unfortunate and something we want to minimize as much as we possibly can, that's a deliberate trade-off (e.g. when we run out of ideas of ways that the browser make things backwards compatible, the next line of change goes towards IDPs - followed by RPs and users).

> This is concerning. This ultimately leads to a world where one company blocks everyone else out but has exclusive access to a superset of data. We recently saw this happen on mobile platforms with platform changes regarding device advertising identifiers.

Yeah, agreed here too, it is not ideal. Two thoughts:

(a) we are generally designing things hoping that one day we would be able to move FedCM to the operating system rather than the browser. i don't know/think this changes much your concern (maybe it just moves it to a lower layer?), but wanted to be transparent about the intent.
(b) i'll have to think a bit harder, but I believe this specific concern, while valid, isn't changing drastically between the pull model and the push model.

**Comment by bvandersloot-mozilla** [comment]
>> I assume the browser owner is aware of any logins I perform

>This is concerning. 

To clarify, I'm not saying this is happening. Rather, that it is possible.

**Comment by gffletch** [comment]
> Yeah, agreed, more specifically, IDPs. While unfortunate and something we want to minimize as much as we possibly can, that's a deliberate trade-off (e.g. when we run out of ideas of ways that the browser make things backwards compatible, the next line of change goes towards IDPs - followed by RPs and users).

I understand the intent but I'm not sure it's true. As we've seen in other issues, there are many additional use cases that FedCM will need to support. As those use cases get added I highly suspect the impact to RPs will go up. There is also much intrenched deployed code that isn't in active development. Since RPs have to change (even if hopefully minimally) fixing that code can be quite challenging and take some time to figure out. We need to be careful to no underestimate the amount of work required to move to FedCM.

**Comment by npm1** [comment]
Just want to note that we addressed the deterministic attacks resulting from the filing of this issue by having a two-tier manifest. We still have the non-deterministic timing attack issue from the credentialed fetch ie https://github.com/fedidcg/FedCM/issues/231

**Comment by samuelgoto** [comment]
For those of you following this thread, I just wanted to report back on the progress that @bvandersloot-mozilla and @cbiesinger have been making. 

Here are a few presentations that we have used to discuss options:

- [Jun 2022](https://github.com/fedidcg/FedCM/blob/main/meetings/2022/FedCM_%20Options%20for%20the%20Timing%20Attack%20Problem%20(8_16_2022).pdf)
- [Aug 2022](https://github.com/fedidcg/FedCM/blob/main/meetings/2022/FedCM_%20Options%20for%20the%20Timing%20Attack%20Problem%20(8_16_2022).pdf)

I think we are arriving at some convergence, which we expect to close at TPAC in person.

**Comment by samuelgoto** [comment]
Just to write down here what we seem to be converging on, it is a variation of what @bvandersloot-mozilla proposed earlier:

> My vision, based on my own priors and what you've said of the push model:
>
> 1.  User logs into the IDP as a first party.
> 1.  IDP calls a browser API to set up the account at the IDP in the browser.
> 1.  Some time later the user presses an RP button that says Log In With X which calls a browser API.
> 1.  The browser searches through its list of accounts that match all X the RP supports
> 1.  The browser presents an account chooser to the user, and includes a link that navigates to the login page of each IDP with no accounts found.
> 1.  User selects an IDP, Auth proceeds using storage, user is logged in.
> 1.  Success

To support step (2), we'd expose a browser API that allows IdPs to tell the browser that the user is logged in (with no UX prompt, a silent store). We've been calling this the "IdP Sign-in Status API", because, well, it is an API that allows IdPs to signal its user's sign-in status. Details largely TBD, but, based on what we heard from IdPs, we think it is likely useful to have a HTTP Header API (for IdPs that use a lot of 302s to log the user in and out) and/or a JS API (for IdPs that actually load content/script to log users in). 

Here is an example of what that could look like:

```http
IdP-Sign-In-Status: action=login
IdP-Sign-In-Status: action=logout
```

And in JS:

```js
// logs the user in
IdentityProvider.login();

// logs the user out
IdentityProvider.logout();
```

To support step (4), the browser then reads its local storage to determine if the user is logged in to the IdP or not.

- If the IdP Sign-in Status API was never called
    - we consider the sign-in status to be `undefined`, and 
    - we fetch the accounts list and
    - set the sign-in status.
- If the user is not logged in
    - we resolve the JS promise immediately: nothing to be done, no timing attack to be paid.
- If the user is logged in
    - we fetch the accounts list
    - if we don't have accounts (e.g. session was invalidated server-side), we show a "login to IdP UI" and then
    - we proceed with step (5).

This variation doesn't fully prevent the timing attack problem, but it seemed like a strict pre-requisite and a good starting point.

**Comment by asr-enid** [comment]
> If we don't have accounts (e.g. session was invalidated server-side)
we show a "login to IdP UI" and then

Am I right that this assumes that an IDP Sign-In UX would be mandatory here (in case a session expired for example), i.e. could not be controlled by the RP but only a users choice to proceed or not? This seems to be the case - see discussion around "silent" parameters in classical flows (prompt=none). 

How would one deal with Multi-IDP Scenarios here? 
- IDP Sign-In in case NONE has a valid active session, if so which (priorized list)
- Offer **Sign-In to RP** directly (if there is at least one signed in account) + **sign-in with other supported IDPs** (for which no sessions are present...)

**Comment by yi-gu** [comment]
> Am I right that this assumes that an IDP Sign-In UX would be mandatory here (in case a session expired for example), i.e. could not be controlled by the RP but only a users choice to proceed or not? 
Yes you are right. The essence of this proposal is to prevent "silent tracking". i.e. whenever a timing attack could happen, we show some UI to users (except for the one-time-per-IdP silent tracking). Given that the API is likely called by the IdP, exposing some control to the RP could actually pass it to the IdP or trackers.

> How would one deal with Multi-IDP Scenarios here?
> 
> * IDP Sign-In in case NONE has a valid active session, if so which (priorized list)
If a user has no active session with any of the IdPs, it's up to the RP to specify the order of IdPs. We are still working on UX details such as whether previous session should count w.r.t. the order. e.g. a user has signed in with one IdP in the past and signed out before the API is called. Suggestions are welcome!

> * Offer **Sign-In to RP** directly (if there is at least one signed in account) + **sign-in with other supported IDPs** (for which no sessions are present...)
This  matches our [intuition](https://github.com/fedidcg/FedCM/blob/main/meetings/2022/FedCM%20%40%20TPAC.pdf) as well. Again, we are still in early UI design stage so suggestions are welcome :)

**Comment by npm1** [comment]
> Reading through the updated [slide deck ](https://github.com/fedidcg/FedCM/blob/main/meetings/2022/FedCM_%20Options%20for%20the%20Timing%20Attack%20Problem%202022-08-31.pdf) it seems this step is meant to not occur with real IDPs albeit edge cases like network errors. It might also happen in case the session is timing out though, in which case the IDP will might not have been able to signal a logout via the API which is also mentioned in the deck

Any insight into how often the session would time out without having an ability to call `logout()` would be useful. I agree that this is potentially a big problem, but I don't have any real IDP data.

**Comment by npm1** [comment]
I think this issue solved with all the work we've done in this area (well-known files and IdP login status)? Let me know if we can close

**Comment by Kithughh** [comment]
This is Kristina Lynn holst as KitKat and I'm deaf as disability, and I believe I am two way head communication with my partner names micheal vickers , in privacy invisible as neurocommuncation as I believed I have someone named Jason skr.... Has a access to my header as controlling my brain in computer, and I can't find the code and closed down this threatening, as I am control by high voltage and volume s and it hurts my head if y'all can find and stop it pls stop this runtime were brain communication in though s possible wave radio communication I live at 329 Indian cir Sparta tenn 38583 pls stop this target

**Comment by Kithughh** [comment]
It starts last year on September 2022 or before pls stop this runtime target

**Comment by Kithughh** [comment]
My email is kitholst77@gmail.com

**Comment by geofft** [comment]
Is eTLD+1 actually sufficient for this? My understanding is that the public suffix list will generally accept any plausible-sounding rationale for private domains. So I can go to them and say, "Hi, I'm launching a cool web hosting platform called eviltracker.example, can I get that on the PSL" and they will probably say "Sure," and then I can set up wildcard DNS and a web server. Then if I tell other origins to make FedCM requests to `https://${some-tracking-string}.eviltracker.example`, I will receive the tracking string in the Host: header.

Put another way, as common as the eTLD+1 terminology is, the security model of the PSL is more eTLD+n for n≥1. The PSL restricts you from setting cookies at the eTLD, but it doesn't prevent you from setting cookies at the eTLD+2, there's no harm in that.  But this proposal seems to rely on eTLD+n for n=1 specifically for security. You can't host the well-known file at the eTLD+2. (And in fact if it were hosted at the eTLD+0 that would be fine, from a security perspective—while it's unrealistic to expect https://com/.well-known/web-identity will exist and list all IDPs on `.com`, if that somehow happened, there wouldn't be a reason to find it untrustworthy. Compare with cookies, which absolutely must not be set on https://com/.) The PSL is geared towards the idea that adding new entries _prevents_ tracking opportunities instead of _creating_ them.

I'm sorry I don't have any great ideas on how to address this. I wonder if it would be workable to insist that the computation of the eTLD+1 is done with solely the "ICANN" entries on the PSL and skip entries on the "PRIVATE" list. The downside is you wouldn't be able to host an IDP on `github.io`, `glitch.me`, etc. But you could always buy a "real" domain name, and maybe most IDPs want to do that anyway.

(I suppose that leaves an attack involving untrustworthy TLD registries tracking you directly, but that one seems much harder to address.)

**Comment by hlflanagan** [comment]
I propose closing this issue for now. If the original poster, any of the commenters, or someone in the WG or CG wants to advocate for keeping it open and help move it forward, please comment within two weeks of this label being applied.

**Comment by hlflanagan** [comment]
Closing. We can re-open this if/when there is someone to champion the discussion.
